NPE in UPDATE with SELECT UNION

Dear H2 Team,

we are testing the latest H2 Git snapshot 2.0.201 and while most things seem to work, we found one query which yields in a NPE as shown below.
This very same query works well with 1.4.200 (GIT from March 2020). So we think this query triggers a regression error.

The query was developed with 1.4.197 where there was no full MERGE support yet. We know that we can rewrite the query as a MERGE statement and we did that already as a work-around.

Thank you for looking at it and cheers.
Andreas

```exception
org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.NullPointerException" [50000-201]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:536)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:459)
	at org.h2.message.DbException.get(DbException.java:189)
	at org.h2.message.DbException.convert(DbException.java:377)
	at org.h2.server.TcpServerThread.sendError(TcpServerThread.java:239)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:190)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.get(Unknown Source)
	at org.h2.schema.Schema.findConstant(Schema.java:442)
	at org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:185)
	at org.h2.expression.ExpressionList.optimize(ExpressionList.java:58)
	at org.h2.expression.Subquery.optimize(Subquery.java:106)
	at org.h2.command.dml.SetClauseList$RowExpression.mapAndOptimize(SetClauseList.java:348)
	at org.h2.command.dml.SetClauseList$SetMultiple.mapAndOptimize(SetClauseList.java:398)
	at org.h2.command.dml.SetClauseList.mapAndOptimize(SetClauseList.java:210)
	at org.h2.command.dml.Update.prepare(Update.java:166)
	at org.h2.command.Parser.prepareCommand(Parser.java:952)
	at org.h2.engine.SessionLocal.prepareLocal(SessionLocal.java:623)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:283)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:188)
	... 1 more

                                     org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.lang.NullPointerException" [50000-201]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:536)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:459)
	at org.h2.message.DbException.get(DbException.java:189)
	at org.h2.message.DbException.convert(DbException.java:377)
	at org.h2.server.TcpServerThread.sendError(TcpServerThread.java:239)q
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:190)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.get(Unknown Source)
	at org.h2.schema.Schema.findConstant(Schema.java:442)
	at org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:185)
	at org.h2.expression.ExpressionList.optimize(ExpressionList.java:58)
	at org.h2.expression.Subquery.optimize(Subquery.java:106)
	at org.h2.command.dml.SetClauseList$RowExpression.mapAndOptimize(SetClauseList.java:348)
	at org.h2.command.dml.SetClauseList$SetMultiple.mapAndOptimize(SetClauseList.java:398)
	at org.h2.command.dml.SetClauseList.mapAndOptimize(SetClauseList.java:210)
	at org.h2.command.dml.Update.prepare(Update.java:166)
	at org.h2.command.Parser.prepareCommand(Parser.java:952)
	at org.h2.engine.SessionLocal.prepareLocal(SessionLocal.java:623)
	at org.h2.server.TcpServerThread.process(TcpServerThread.java:283)
	at org.h2.server.TcpServerThread.run(TcpServerThread.java:188)
	... 1 more
```


```sql
update cfe.impairment b
set (b.impairment_stage, b.default_date, b.overdue_days, b.risk_classification) = 
	(SELECT 
		   3 impairment_stage
		   , a.overdue_since default_date
		   , overdue_days
		   , '90 DPD' risk_classification
	FROM   cfe.EXT_OVERDUE_INFORMATION a
	WHERE  a.value_date = To_date('9/11/20', 'mm/dd/yy')
		   AND a.foracid=b.id_instrument
		   AND a.overdue_bucket IN ( 'D' )
	UNION ALL
	SELECT 2               impairment_stage
		   , null            default_date
		   , overdue_days
		   , '30 DPD' risk_classification
	FROM   cfe.EXT_OVERDUE_INFORMATION a
	WHERE  a.value_date = To_date('9/11/20', 'mm/dd/yy')
		   AND a.foracid=b.id_instrument
		   AND a.overdue_bucket IN ( '60', '90' ) 
	UNION ALL
	SELECT 1               impairment_stage
		   , null            default_date
		   , overdue_days
		   , null           risk_classification
	FROM   cfe.EXT_OVERDUE_INFORMATION a
	WHERE  a.value_date = To_date('9/11/20', 'mm/dd/yy')
		   AND a.foracid=b.id_instrument
		   AND a.overdue_bucket IN ( '30' ))
where exists 
	(select 1 from cfe.EXT_OVERDUE_INFORMATION a
		WHERE  a.value_date = To_date('9/11/20', 'mm/dd/yy')
		AND a.foracid=b.id_instrument)
```

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,10 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2882: NPE in UPDATE with SELECT UNION
+</li>
+<li>PR #2881: Store users and roles together and user-defined functions and aggregates together
+</li>
 <li>Issue #2878: Disallow spatial indexes in PageStore databases
 </li>
 <li>PR #2874: Use 64-bit row counts in results and other changes

--- a/h2/src/main/org/h2/expression/ExpressionList.java
+++ b/h2/src/main/org/h2/expression/ExpressionList.java
@@ -61,12 +61,16 @@ public final class ExpressionList extends Expression {
             }
             list[i] = e;
         }
-        type = isArray ? TypeInfo.getTypeInfo(Value.ARRAY, list.length, 0, TypeInfo.getHigherType(list))
-                : TypeInfo.getTypeInfo(Value.ROW, 0, 0, new ExtTypeInfoRow(list));
+        initializeType();
         if (allConst) {
             return ValueExpression.get(getValue(session));
         }
         return this;
+    }
+
+    void initializeType() {
+        type = isArray ? TypeInfo.getTypeInfo(Value.ARRAY, list.length, 0, TypeInfo.getHigherType(list))
+                : TypeInfo.getTypeInfo(Value.ROW, 0, 0, new ExtTypeInfoRow(list));
     }
 
     @Override

--- a/h2/src/main/org/h2/expression/Subquery.java
+++ b/h2/src/main/org/h2/expression/Subquery.java
@@ -103,7 +103,9 @@ public final class Subquery extends Expression {
             for (int i = 0; i < columnCount; i++) {
                 list[i] = expressions.get(i);
             }
-            expression = new ExpressionList(list, false).optimize(session);
+            ExpressionList expressionList = new ExpressionList(list, false);
+            expressionList.initializeType();
+            expression = expressionList;
         }
         return this;
     }

--- a/h2/src/main/org/h2/jdbc/meta/DatabaseMetaLocal.java
+++ b/h2/src/main/org/h2/jdbc/meta/DatabaseMetaLocal.java
@@ -236,9 +236,10 @@ public final class DatabaseMetaLocal extends DatabaseMetaLocalBase {
                     }
                     for (int i = 0; i < methods.length; i++) {
                         JavaMethod method = methods[i];
+                        TypeInfo typeInfo = method.getDataType();
                         getProceduresAdd(result, catalogValue, schemaValue, procedureNameValue,
                                 userDefinedFunction.getComment(),
-                                method.getDataType().getValueType() != Value.NULL ? PROCEDURE_RETURNS_RESULT
+                                typeInfo == null || typeInfo.getValueType() != Value.NULL ? PROCEDURE_RETURNS_RESULT
                                         : PROCEDURE_NO_RESULT,
                                 getString(procedureName + '_' + (i + 1)));
                     }
@@ -327,10 +328,10 @@ public final class DatabaseMetaLocal extends DatabaseMetaLocalBase {
                 for (int i = 0, l = methods.length; i < l; i++) {
                     JavaMethod method = methods[i];
                     Value specificNameValue = getString(procedureName + '_' + (i + 1));
-                    TypeInfo type = method.getDataType();
-                    if (type.getValueType() != Value.NULL) {
+                    TypeInfo typeInfo = method.getDataType();
+                    if (typeInfo != null && typeInfo.getValueType() != Value.NULL) {
                         getProcedureColumnAdd(result, catalogValue, schemaValue, procedureNameValue, specificNameValue,
-                                type, method.getClass().isPrimitive(), 0);
+                                typeInfo, method.getClass().isPrimitive(), 0);
                     }
                     Class<?>[] columnList = method.getColumnClasses();
                     for (int o = 1, p = method.hasConnectionParam() ? 1 : 0, n = columnList.length; p < n; o++, p++) {

--- a/h2/src/main/org/h2/table/InformationSchemaTable.java
+++ b/h2/src/main/org/h2/table/InformationSchemaTable.java
@@ -1969,7 +1969,7 @@ public final class InformationSchemaTable extends MetaTable {
                         FunctionAlias.JavaMethod method = methods[i];
                         TypeInfo typeInfo = method.getDataType();
                         String routineType;
-                        if (typeInfo.getValueType() == Value.NULL) {
+                        if (typeInfo != null && typeInfo.getValueType() == Value.NULL) {
                             routineType = "PROCEDURE";
                             typeInfo = null;
                         } else {

--- a/h2/src/main/org/h2/table/InformationSchemaTableLegacy.java
+++ b/h2/src/main/org/h2/table/InformationSchemaTableLegacy.java
@@ -1381,6 +1381,9 @@ public final class InformationSchemaTableLegacy extends MetaTable {
                         }
                         for (FunctionAlias.JavaMethod method : methods) {
                             TypeInfo typeInfo = method.getDataType();
+                            if (typeInfo == null) {
+                                typeInfo = TypeInfo.TYPE_NULL;
+                            }
                             add(session,
                                     rows,
                                     // ALIAS_CATALOG
@@ -1459,7 +1462,7 @@ public final class InformationSchemaTableLegacy extends MetaTable {
                         for (FunctionAlias.JavaMethod method : methods) {
                             // Add return column index 0
                             TypeInfo typeInfo = method.getDataType();
-                            if (typeInfo.getValueType() != Value.NULL) {
+                            if (typeInfo != null && typeInfo.getValueType() != Value.NULL) {
                                 DataType dt = DataType.getDataType(typeInfo.getValueType());
                                 add(session,
                                         rows,

--- a/h2/src/test/org/h2/test/scripts/dml/update.sql
+++ b/h2/src/test/org/h2/test/scripts/dml/update.sql
@@ -213,3 +213,25 @@ DROP TABLE TEST;
 
 DROP DOMAIN D;
 > ok
+
+CREATE TABLE TEST(A INT, B INT, C INT) AS VALUES (0, 0, 1), (0, 0, 3);
+> ok
+
+CREATE TABLE S1(A INT, B INT) AS VALUES (1, 2);
+> ok
+
+CREATE TABLE S2(A INT, B INT) AS VALUES (3, 4);
+> ok
+
+UPDATE TEST SET (A, B) = (SELECT * FROM S1 WHERE C = A UNION SELECT * FROM S2 WHERE C = A);
+> update count: 2
+
+TABLE TEST;
+> A B C
+> - - -
+> 1 2 1
+> 3 4 3
+> rows: 2
+
+DROP TABLE TEST, S1, S2;
+> ok

